### PR TITLE
Add pattern to merge/simplify split-cat

### PR DESCRIPTION
**Pull request by devashishshankar**
Summary:
In simple cases, both split and cat node can be removed in a "split->cat" pattern. However, there are various cases where they can't simply be removed and we need to simplify split/ add transforms before cat. Some such cases are:
* Split-dim != cat-dim (but equal split)
* Final node: cat vs stack
* Final node has additional args
* Shuffling of args between split/cat
* Some final nodes are non-(cat/stack)

For more details, please refer to https://docs.google.com/presentation/d/1SxBuY_FZfljSlX6i8slRNgP2CsUCICP0o4qe8cNNX8U/edit#slide=id.g232e9a90f64_0_273 (slides 8-15)

Differential Revision: D45452404

cc @soumith @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

### CODE REVIEW

File: torch/_inductor/fx_passes/split_cat.py [hunk 1]
```diff
@@ -189,9 +192,385 @@ def merge_splits(
...
+        user_inputs_list_new = self.replace_split(
+            graph, split_node, user_inputs_list, simplified_split_ranges
+        )
+        self.replace_cat(
+            graph, split_node, next_users, user_inputs_list_new, transform_params_list
+        )
+        self.erase_old_nodes(graph, split_node, next_users)
+
+    def find_next_users(self, split_node):
+        next_users = set()
```
Comment (jansel) [inline torch/_inductor/fx_passes/split_cat.py:243]: Sets are not determinstic in Python.

**Review by jansel — CHANGES_REQUESTED** [review]
This is done in a pretty ad hoc way.

1) Can we break it up into smaller steps?
2) Can we use more of the pattern matcher infra?  I worry there will be subtle issues hiding in places.

File: torch/_inductor/fx_passes/split_cat.py (hunk 1 above)
Comment (devashishshankar) [reply torch/_inductor/fx_passes/split_cat.py:243]: The order of `next_users` doesn't matter. `next_users` are essentially things like `cat`, `stack` nodes (not getitems), so we can replace them in any order

File: torch/_inductor/fx_passes/split_cat.py [hunk 2]
```diff
@@ -215,3 +594,16 @@ def _split_cat_init():
...
+            torch.cat,
+        ),
+        CallFunctionVarArgs(
+            torch.stack,
+        ),
+    ]:
+        pattern = MergeSplitCat(
+            pattern=pattern, extra_check=config_flag("split_cat_fx_passes")
+        )
+        pattern.register(merge_split_cat_pass)
```
[the remark this answers is not in the document]
Comment (devashishshankar) [reply torch/_inductor/fx_passes/split_cat.py:609]: That's a simple change that I can make. Although I wanted a class structure, since there are so many functions - class helps group all related functions together. And I'm not sure if the decorator would work on a class method.

File: torch/_inductor/fx_passes/split_cat.py (hunk 1 above)
Comment (jansel) [reply torch/_inductor/fx_passes/split_cat.py:243]: Replacing them in random order will make the output code of `torch.compile()` non-determinstic.  We consider that a bug.

**Review by jansel — APPROVED** [review]
